src/PrestaShopBundle/Translation/Provider/ModulesProvider.php

```diff
@@ -24,43 +24,316 @@
+use PrestaShop\TranslationToolsBundle\Translation\Helper\DomainHelper;
+use PrestaShopBundle\Translation\DomainNormalizer;
+use PrestaShopBundle\Translation\Exception\UnsupportedLocaleException;
+use PrestaShopBundle\Translation\Extractor\LegacyModuleExtractorInterface;
+use PrestaShopBundle\Translation\Loader\DatabaseTranslationReader;
+use PrestaShopBundle\Translation\Provider\Catalogue\DefaultCatalogueProvider;
+use PrestaShopBundle\Translation\Provider\Catalogue\FileTranslatedCatalogueProvider;
+use PrestaShopBundle\Translation\Provider\Catalogue\TranslationCatalogueProviderInterface;
+use PrestaShopBundle\Translation\Provider\Catalogue\UserTranslatedCatalogueProvider;
+use Symfony\Component\Translation\Loader\LoaderInterface;
+use Symfony\Component\Translation\MessageCatalogue;
+use Symfony\Component\Translation\MessageCatalogueInterface;
+
 /**
- * Translation provider for native modules (maintained by the core team)
- * Translations are provided by Crowdin.
+ * Retrieves translations from modules
  */
-class ModulesProvider extends AbstractProvider implements UseDefaultCatalogueInterface
+class ModulesProvider implements ProviderInterface
 {
     /**
-     * {@inheritdoc}
+     * @var string Path where translation files are found
+     */
+    protected $modulesDirectory;
+
+    /**
+     * @var string
+     */
+    private $translationsDirectory;
+
+    /**
+     * @var string domain
+     */
+    protected $domain;
+
+    /**
+     * @var LoaderInterface Translation loader from legacy files
+     */
+    private $legacyFileLoader;
+
+    /**
+     * @var LegacyModuleExtractorInterface Extractor
+     */
+    private $legacyModuleExtractor;
+
+    /**
+     * @var string Module name
+     */
+    private $moduleName;
+
+    /**
+     * @var MessageCatalogueInterface[]
      */
-    public function getTranslationDomains()
+    private $defaultCatalogueCache;
+
+    /**
+     * @var DatabaseTranslationReader
+     */
+    private $databaseReader;
+
+    /**
+     * @param DatabaseTranslationReader $databaseReader
+     * @param string $modulesDirectory
+     * @param string $translationsDirectory
+     * @param LoaderInterface $legacyFileLoader
+     * @param LegacyModuleExtractorInterface $legacyModuleExtractor
+     * @param string $moduleName
+     */
+    public function __construct(
+        DatabaseTranslationReader $databaseReader,
+        string $modulesDirectory,
+        string $translationsDirectory,
+        LoaderInterface $legacyFileLoader,
+        LegacyModuleExtractorInterface $legacyModuleExtractor,
+        string $moduleName
+    ) {
+        $this->legacyFileLoader = $legacyFileLoader;
+        $this->legacyModuleExtractor = $legacyModuleExtractor;
+        $this->databaseReader = $databaseReader;
+        $this->modulesDirectory = $modulesDirectory;
+        $this->translationsDirectory = $translationsDirectory;
+        $this->moduleName = $moduleName;
+    }
+
+    /**
+     * @param string $locale
```

Comment:
Suggestion: explain the fallback mechanism implemented in it in phpDoc

Reply:
@sowbiba